### PR TITLE
[DRAFT] Workflows Docs

### PR DESCRIPTION
**Pull request by buggyhunter**

### CODE REVIEW

File: qstash/workflows/getstarted.mdx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+- It provides **all the power of QStash** while abstracting away the complexities, allowing you to focus on implementing the **essential parts**.
+
+Here is an example how an an endpoint might use our SDK:
+
+```javascript 
+import { serve } from "@upstash/qstash/workflow/nextjs";
+import { sendEmail, UserQuestionRequest } from "./utils"
+
+export const POST = serve<{userRequest: UserQuestionRequest}>({async context => {
+    await context.sleep("sleepDays" , 3d)
```
Comment (mehmettokgoz) [inline qstash/workflows/getstarted.mdx:19]: At first glance, the logic of immediately sleeping for 3 days doesn’t make sense to me. A more logical flow might be:

1.	Save the user to the database.
2.	Sleep for the specified duration.
3.	Make the external call and send the email.

WDYT?

Also, for the sleep durations, I noticed you’re using invalid inputs like `3d` or utility functions. Would it be clearer to use seconds directly, like `3 * 24 * 60 * 60`? Since people might copy-paste these code snippets as templates, it would be better to make them directly usable. 

```suggestion
    await context.run("saveUser", async() => {
        await saveUser(userRequest)
    })
    
    await context.sleep("sleepDays" , 3d)
```

File: qstash/workflows/basics/api.mdx [hunk 2]
```diff
@@ -0,0 +1,119 @@
+---
+title: "API"
+---
+
+# context.run
+
+`context.run` requires a step name and the code of the step as an async function. Each step can run for the full duration allowed by the platform. The result of each step is stored and remembered by QStash, allowing it to continue from where it left off in case of failure (e.g., endpoint downtime).
+
+Here are a couple of examples to demonstrate its usage:
```
Comment (mehmettokgoz) [inline qstash/workflows/basics/api.mdx:9]: Some questions from the user’s POV:

What happens if my async function throws an exception?
Do I need to handle it manually, or will QStash automatically retry the function?
If it retries automatically, how can I configure the retry behavior; e.g. max number of retries?

File: qstash/workflows/basics/api.mdx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+    await Promise.all([promise1, promise2]);
+  },
+});
+```
+</CodeGroup>
+
+# context.sleep
+
+`context.sleep` postpones the execution of the code for a specified duration. The code will exit and not consume any function time. After the specified duration, the endpoint will be called again to resume execution.
```
Comment (mehmettokgoz) [inline qstash/workflows/basics/api.mdx:47]: It’s better to emphasize that the duration is in seconds here.

File: qstash/workflows/basics/api.mdx (hunk 2 above)
Comment (CahidArda) [reply qstash/workflows/basics/api.mdx:9]: retry amount can't be configured at this moment. I was thinking of adding it in a later release.

I can continue working on it if you think it's a must.

File: qstash/workflows/getstarted.mdx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    )
+
+    await context.run("sendEmail", async() => {
+        await sendEmail(userRequest, response)
+    })
+  }
+})
+```
+
+This example is for nextjs, to check all other plaforms and usage examples see [Supported Platforms](/qstash/workflows/platforms/platforms)
```
Comment (mehmettokgoz) [inline qstash/workflows/getstarted.mdx:37]:
```suggestion
This example is for Next.js, to check all other plaforms and usage examples see [Supported Platforms](/qstash/workflows/platforms/platforms)
```

File: qstash/workflows/basics/api.mdx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    });
+  },
+});
+```
+
+# context.call
+
+`context.call` is used to call an HTTP endpoint as a step. While this can be done inside `context.run`, the key difference is that `context.call` does not wait for the endpoint to return. This allows you to bypass the time limitations of the platform. When the response is ready, the function will continue as if resuming from the return of the `context.call`.
+
+The endpoint can take up to 15 minutes or 2 hours to respond, depending on your [QStash plan(Max HTTP Connection Timeout)](https://upstash.com/pricing/qstash)T.
```
Comment (mehmettokgoz) [inline qstash/workflows/basics/api.mdx:99]: User's POV:

How can I configure timeout duration for my external call?

File: qstash/workflows/basics/api.mdx (hunk 2 above)
Comment (mehmettokgoz) [reply qstash/workflows/basics/api.mdx:9]: I think it's OK to postpone it for now.

My point is to put the answers in the documentation, even if we don’t support them.

**Comment by mehmettokgoz** [comment]
A dedicated page (or section) about local development would be valuable in the `How To`.

I’m trying to run my workflow locally, but I’m getting various error messages, like missing `Upstash-Signature` header. I think we should explain the flow for users.

Or is this something we plan to explain in the platforms section?

WDYT? @sancar @CahidArda

File: qstash/workflows/basics/api.mdx (hunk 3 above)
Comment (sancar) [reply qstash/workflows/basics/api.mdx:47]: @CahidArda Can we have a better type to give things like 3d(for 3 days) , 1h(for 1 hour)

File: qstash/workflows/basics/api.mdx (hunk 5 above)
Comment (sancar) [reply qstash/workflows/basics/api.mdx:99]: If I document now, it will be a bit weird since feature is not there. I think we can add the doc after adding the feature. Right now, the limit is the plan as it is also understood from the doc.

File: qstash/workflows/basics/api.mdx (hunk 2 above)
Comment (sancar) [reply qstash/workflows/basics/api.mdx:9]: I will mention that the default is 3 and configuration will come soon in the doc.

File: qstash/workflows/basics/caveats.mdx [hunk 6]
```diff
@@ -0,0 +1,174 @@
+---
+title: "Caveats"
+---
```
[the remark this answers is not in the document]
Comment (sancar) [reply qstash/workflows/basics/caveats.mdx:3]: There are no concurrency limit now, so I prefer not to mention anything right now. 
When we introduce queue, the limits will be there. We can mention it then. 

I will add a section for nested API calls

**Comment by sancar** [comment]
@mehmettokgoz Do you mean something like this right ? Using a a tunnel service ?
https://upstash-workflows.mintlify.app/qstash/howto/local-tunnel
I think we can add something similar for workflows as well. The content will be almost same. 
I think we can add it under `How To` as `Local Development`

**Comment by CahidArda** [comment]
we discussed local development in detail this morning as the dx team. Here is what we have in mind:

There will be a script to start an ngrok local tunnel:
```
npx @upstash/workflow-cli start
```
It will start a local tunnel and export its url as an environment variable `UPSTASH_WORKFLOW_URL`.

Whenever a request arrives, the sdk will check if the url has localhost in it and `UPSTASH_WORKFLOW_URL` env var is set, localhost in the url will be replaced with the `UPSTASH_WORKFLOW_URL` value.

I can write the local dev section when we finish this.

**Comment by sancar** [comment]
@CahidArda I have added local development section with ngrok. You can update it as you see if after your suggested improvement.

File: qstash/workflows/basics/api.mdx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    await context.run("sendWelcomeEmail", async () => {
+      return sendEmail(user.name, user.email);
+    });
+  },
+});
+```
+
+# context.sleepUntil
+
+`context.sleepUntil` postpones the execution of the code until a specified date. The code will exit and not consume any function time. On the given date, the endpoint will be called again to resume execution.
```
Comment (mehmettokgoz) [inline qstash/workflows/basics/api.mdx:74]: Is it worth mentioning that they need to `await` sleep calls? Otherwise, their flow might not work as expected.

A user might naively assume that execution will pause automatically (I'm not sure though). WDYT?

File: qstash/workflows/howto/failures.mdx [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+The `failureFunction` allows you to define custom logic for handling failures. 
+
+#### Example Configuration:
+
+```javascript
+export const POST = serve<string>(
+  failureFunction: async (context, failStatus, failResponse) => {
+    const input = context.requestPayload;
+    cleanupDB(input)
+  },
```
Comment (CahidArda) [inline qstash/workflows/howto/failures.mdx:35]: having the `context` parameter is really complex right now. It's partly made complex because of the auth-before-all-steps approach we have.

Here is the current parameters:
```
    status: number,
    header: Record<string, string>,
    body: FailureFunctionPayload,
    workflowRunId: string
```

Removing our auth approach, adding an onInvocation callback for authentication and adding context parameter to  failureFunction all combined is simpler than adding context to failureFunction in the current setup :D

TODO: parameter order is wrong too.

File: qstash/workflows/howto/cancel.mdx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+You can cancel a workflow run both programatically and from console. 
+
+## Console 
+
+From the [Workflows tab](https://console.upstash.com/qstash?tab=workflows), after finding the related workflow you can cancel the workflow run as follows:
+
+<Frame>
+  <img src="/img/qstash/qstash-workflows-cancel.png" />
+</Frame>
```
Comment (CahidArda) [inline qstash/workflows/howto/cancel.mdx:13]: screenshot is outdated

File: qstash/workflows/howto/schedule.mdx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+    const userSign = await context.run("registerUser", async () => {
+      return registerUser(input);
+    });
+
+    if (userSign.success) {
+      await context.run("triggerSchedule", async () => {
+        await client.schedules.create({
+          scheduleId: "newsSchedule:" + userSign.email,
+          destination: "https://cool-app.vercel.app/sendNewsletter",
```
Comment (CahidArda) [inline qstash/workflows/howto/schedule.mdx:113]: did we try the scheduleId parameter as in here? A user on intercom mentioned that they used it too even though I only recently created a PR for it

File: qstash/workflows/basics/caveats.mdx (hunk 6 above)
[the remark this answers is not in the document]
Comment (CahidArda) [reply qstash/workflows/basics/caveats.mdx:3]: what is a nested API call?

[the remark this answers is not in the document]
Comment (mehmettokgoz) [reply qstash/workflows/basics/caveats.mdx:3]: Nested `context.run` or `context.sleep` calls like following:

```
context.run("step", async () => {
    // This is not allowed
    await context.sleepUntil("sleep", 100)

    // This is not allowed
    await context.run("nested run", async () => {
    })
})
```

[the remark this answers is not in the document]
Comment (CahidArda) [reply qstash/workflows/basics/caveats.mdx:3]: This was added I think so I will resolve the thread 👍

The sdk will throw if this happens, we specify it to be undefined behavior which is fine too I think.

tests https://github.com/upstash/qstash-js/blob/DX-999/src/client/workflow/context.test.ts#L14

File: qstash/workflows/getstarted.mdx [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    await context.sleep("sleepDays" , 3d)
+
+    await response = context.call(
+      "call open ai",
+      "https://api.openai.com/v1/chat/completions",
+      "POST",  "Authorization: Bearer $OPENAI_API_KEY",
+      {"model": "gpt-3.5-turbo-16k",
+        "messages": [{"role": "system", "content": "You are a helpful assistant." },
+                     {"role": "user","content": request.question}]}
+    )
```
[the remark this answers is not in the document]
Comment (sancar) [reply qstash/workflows/getstarted.mdx:32]: I did not. We should not be doing base64 encode but I will try.

File: qstash/workflows/howto/schedule.mdx (hunk 10 above)
Comment (sancar) [reply qstash/workflows/howto/schedule.mdx:113]: I added this example assuming that we will have it soon.

File: qstash/workflows/platforms/nuxt.mdx [hunk 12]
```diff
@@ -2,61 +2,134 @@
...
+     async (context) => {
+       // Your workflow steps
+     },
+     {
+       // Conditionally set in development, but not in production
+       // as the baseUrl is automatically set in production
+       baseUrl:
+         process.env.NODE_ENV === "development"
+           ? "https://<YOUR_NGROK_OR_TUNNEL_URL>/"
+           : undefined,
```
Comment (CahidArda) [inline qstash/workflows/platforms/nuxt.mdx]: baseUrl is not set in production.

I think it's better if we use UPSTASH_WORKFLOW_URL in development and not set it at all in production @joschan21 

We can also remove "Remove Local Development Settings" from Step 5 if we use UPSTASH_WORKFLOW_URL. We can just say they can omit UPSTASH_WORKFLOW_URL in "Environment Variables"

Same goes for other platforms, cf & next

File: qstash/workflow/quickstarts/cloudflare-workers.mdx [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+interface Env {
+  ENVIRONMENT: "development" | "production"
+}
+
+export default {
+  fetch(req: Request, env: Env, ctx: ExecutionContext) {
+    // 👇 For TypeScript users: define initial payload type
+    return serve<{ text: string }>(
+      async (context) => {
```
Comment (CahidArda) [inline qstash/workflow/quickstarts/cloudflare-workers.mdx:60]: this example doesn't look correct. note that we don't do anything with req/env/ctx variables.

It should have been like [this](https://github.com/upstash/docs/pull/238/commits/5af4011063a0270cd4153392d221335bdfc68013#diff-f0bab82153766e46e18e860e3acdb9486170693c3a9f658cd6da83b1c3bc95d1R65)

File: qstash/workflow/quickstarts/nuxt.mdx [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+To define a workflow endpoint in a Nuxt.js project, navigate into the Nuxt.js `server/api` directory and create a new file, for example called `workflow.ts`. This file will contain your workflow:
+
+```typescript server/api/workflow.ts
+import { serve } from "@upstash/qstash/nuxt"
+
+export const POST = serve(
+  async (context) => {
+    await context.run("initial-step", () => {
+      console.log("initial step ran")
+    })
```
Comment (CahidArda) [inline qstash/workflow/quickstarts/nuxt.mdx:48]:
```javascript
import { serve } from "@upstash/qstash/h3";

export default serve<string>(async (context) => {
  ...
})
```

File: qstash/workflows/platforms/nuxt.mdx (hunk 12 above)
Comment (joschan21) [reply qstash/workflows/platforms/nuxt.mdx]: got it and agree, addressed in following commit

File: qstash/workflow/quickstarts/cloudflare-workers.mdx (hunk 13 above)
Comment (joschan21) [reply qstash/workflow/quickstarts/cloudflare-workers.mdx:60]: this example comes straight from the CF workers boilerplate, what about it is not correct?

File: qstash/workflow/quickstarts/nuxt.mdx (hunk 14 above)
Comment (joschan21) [reply qstash/workflow/quickstarts/nuxt.mdx:48]: good catch, fixed

File: qstash/workflow/quickstarts/cloudflare-workers.mdx (hunk 13 above)
Comment (CahidArda) [reply qstash/workflow/quickstarts/cloudflare-workers.mdx]: [serve method of cloudflare handles the req/env/ctx parameters in it](https://github.com/upstash/qstash-js/blob/main/platforms/cloudflare.ts#L64). This way, you can basically say and let serve do the work:
```javascript
export default {
  fetch: serve<{text: string}>(async context => {
    // ...
  }
}
```

see our example [here](https://github.com/upstash/qstash-js/blob/main/examples/workflow/cloudflare-workers/src/index.ts)

File: qstash/workflow/quickstarts/vercel-nextjs.mdx [hunk 15]
```diff
@@ -60,6 +62,23 @@ export const POST = serve(
...
+```typescript app/api/workflow/route.ts
+
+import { serve } from "@upstash/qstash/nextjs"
+
+export const POST = (req: Request) => {
+  // do something with native request object
+
+  return serve(async (context) => { 
+    // Your workflow steps
+   })
```
Comment (CahidArda) [inline qstash/workflow/quickstarts/vercel-nextjs.mdx:77]: this won't work as the serve method is not passed the req at all.

should have been like this:
```javascript
export const POST = (req: Request) => {
  // do something with native request object

  const handler = serve(async (context) => { 
    // Your workflow steps
   })
  return await handler(req)
}

```

Comment (CahidArda) [reply qstash/workflow/quickstarts/vercel-nextjs.mdx:78]: resolving this thread as it looks like this code snippet is removed

File: qstash/workflow/quickstarts/cloudflare-workers.mdx (hunk 13 above)
Comment (CahidArda) [reply qstash/workflow/quickstarts/cloudflare-workers.mdx]: this issue remains in three cf examples @joschan21

File: qstash/workflow/quickstarts/vercel-nextjs.mdx [hunk 16]
```diff
@@ -84,35 +85,9 @@ export const POST = (req: Request) => {
 
 For QStash, a tool that's used to run workflows under the hood, you need to provide a live URL for your workflow endpoint. Here's how to [set up your workflow endpoint for local development](qstash/workflow/howto/local-development).
 
-Once you have a live URL, proceeed with either one of the following steps:
+In production, a workflow's `baseUrl` is automatically set and can be omitted.
 
-### Setting an enviroment variable (recommended)
```
Comment (joschan21) [inline qstash/workflow/quickstarts/vercel-nextjs.mdx]: Why did you remove this part?
